src/Chainweb/BlockHeaderDB/PruneForks.hs

```diff
@@ -41,180 +55,199 @@ import System.LogLevel
-        streamToHashSet_
-
-    -- Bloom filter for marking block headers
-    --
-    let s = max (int rootHeight + 10 * HS.size roots) (int rootHeight * 2)
-    let (size, hashNum) = BF.suggestSizing s 0.0001
-    !marked <- stToIO $ BF.new (BF.cheapHashes hashNum) size
+pruneForks_ logg _ (MaxRank (Max mar)) _  _
+    | mar <= 1 = 0 <$ logg Warn ("Skipping database prunning for max bound of " <> sshow mar)
+pruneForks_ logg cdb mar mir callback = do
+    logg Debug $ "Prunning block header database for chain " <> sshow (_chainId cdb)
+        <> " with upper bound " <> sshow (_getMaxRank mar)
+        <> " and lower bound " <> sshow (_getMinRank mir)
+
+    !pivots <- entries cdb Nothing Nothing (Just $ MinRank $ Min $ _getMaxRank mar) (Just mar)
+        $ fmap (force . L.nub) . S.toList_ . S.map _blockParent
+            -- the set of pivots is expected to be very small. In fact it is
+            -- almost always a singleton set.
+
+    if null pivots
+      then do
+        logg Warn
+            $ "Skipping database pruning because of an empty set of block headers at upper pruning bound " <> sshow mar
+        return 0
+      else
+        withReverseHeaderStream cdb (mar - 1) mir
+            $ S.foldM_ go (return (pivots, int (_getMaxRank mar), 0)) (\(_,_,!n) -> evaluate n)
 
-    -- Bloom filter for marking block payload hashes
-    --
-    -- Payloads can be shared between different blocks. Thus, to be on the safe
-    -- side, we are not deleting any marked payload.
-    --
-    -- Note that we could use cuckoo hashes or counting bloom filters to do
-    -- reference counting and collect more unreferenced payloads.
-    --
-    -- TODO: would it be better to use a single shared filter?
-    --
-    let (psize, phashNum) = BF.suggestSizing s 0.0001
-    !markedPayloads <- stToIO $ BF.new (BF.cheapHashes phashNum) psize
+  where
+    go :: ([BlockHash], BlockHeight, Int) -> BlockHeader -> IO ([BlockHash], BlockHeight, Int)
+    go ([], _, _) cur = throwM $ InternalInvariantViolation
+        $ "PrunForks.pruneForks_: no pivots left at block " <> encodeToText (ObjectEncoded cur)
+    go (!pivots, !prevHeight, !n) !cur
+        -- This checks the consistency of the database.
+        | prevHeight /= curHeight && prevHeight /= curHeight + 1 =
+            throwM $ InternalInvariantViolation
+                $ "PrunForks.pruneForks_: detected a corrupted database. Some block headers are missing"
+                <> ". Current pivots: " <> encodeToText pivots
+                <> ". Current header: " <> encodeToText (ObjectEncoded cur)
+                <> ". Previous height: " <> sshow prevHeight
+        | _blockHash cur `elem` pivots = do
+            callback False cur
+            let !pivots' = force $ L.nub $ _blockParent cur : L.delete (_blockHash cur) pivots
+            return (pivots', curHeight, n)
+        | otherwise = do
+            deleteHdr cur
+            callback True cur
+            return (pivots, curHeight, n+1)
```

Comment:
What is the loop invariant here, since only `n` is changed, which is not read in `go`?

Reply:
`n` counts the number of deleted blocks. The pivots are unchanged.

The algorithm works as follows:

1. initialization: pivots are initialized with the parents hashs of all blocks that should be kept. All ancestors of these blocks are kept. All other blocks are deleted.

2. traverse all blocks in the db reverse ordered by block height. For each block check:

   1. If the block hash is in the pivot set it's an ancestor of a block that is kept and not deleted. The respective pivot is replaced with the the current block hash.

   2. If the block hash is not in the pivot set, it's not an ancestor of a block that is kept. The block is deleted and the count of deleted blocks `n` is increased by 1.

  There is an additional (not strictly needed) check, that ensures that actually a match for each pivot is found. This ensures that the database was consistent in first place.

Reply:
Right, and I forgot that since it's a `foldM_` the loop invariant is guaranteed, sorry. Thanks for the explanation